pkg/workflow/providers/oam/apply.go

```diff
@@ -117,6 +129,54 @@ func (p *provider) ApplyComponent(ctx wfContext.Context, v *value.Value, act wfT
 	return nil
 }
 
+// ApplyComponent apply component.
+func (p *provider) ApplyComponent(ctx wfContext.Context, v *value.Value, act wfTypes.Action) error {
+	return p.applyComponent(ctx, v, act, nil)
+}
+
+// ApplyComponents apply components in parallel.
+func (p *provider) ApplyComponents(ctx wfContext.Context, v *value.Value, act wfTypes.Action) error {
+	components, err := v.LookupValue("components")
+	if err != nil {
+		return err
+	}
+	parallelism, err := v.GetInt64("parallelism")
+	if err != nil {
+		return err
+	}
+	if parallelism <= 0 {
+		return errors.Errorf("parallelism cannot be smaller than 1")
+	}
+	mu := &sync.Mutex{}
+	var wg sync.WaitGroup
+	ch := make(chan struct{}, parallelism)
```

Comment:
why we need using both this `channel` and `waitGroup`

Reply:
1. In my view, we don't need to limmit the concurrency in the control plane, this won't be large throughput. You can leave it to the go runtime.
2. If you really want to control the parallelism，you can just use a channel, like below:

```
// ApplyComponents apply components in parallel.
func (p *provider) ApplyComponents(ctx wfContext.Context, v *value.Value, act wfTypes.Action) error {
	components, err := v.LookupValue("components")
	if err != nil {
		return err
	}
	parallelism, err := v.GetInt64("parallelism")
	if err != nil {
		return err
	}
	if parallelism <= 0 {
		return errors.Errorf("parallelism cannot be smaller than 1")
	}
	mu := &sync.Mutex{}
	type sig struct {
		err error
		msg string
	}
	ch := make(chan sig, parallelism+1)
	go func() {
		err = components.StepByFields(func(name string, in *value.Value) (bool, error) {
			go func(_name string, _in *value.Value) {
				subErr := p.applyComponent(ctx, _in, act, mu)
				ch <- sig{
					err: subErr,
					msg: "component: "+ _name,
				}
			}(name, in)
			return false, nil
		})
		if err != nil {
			ch <- sig{
				err: err,
				msg: "failed to looping over components",
			}
		} else {
			ch <- sig{msg: "main"}
		}
	}()
for _,sig:=range ch{
// collect error and wait for signal here	
}
	return nil
}
```

Reply:
There will be another PR to handle the concurrent execution framework problem.